### ISSUE TITLE
[RFC] Implement a unique numerical index for Alpaka devices ?

### ISSUE THREAD

**Issue by fwyzard** [alpaka-group/alpaka#1927]
This is a bit complicated to explain.

Some parts of the CMS code that were originally written in CUDA rely on having a unique, progressive numerical index associated to each device.

After the migration to Alpaka, those parts still work fine for the CPU, CUDA and HIP backends: a CPU device always has index `0`, while the CUDA/HIP devices use the underlying index (for example, as returned by `getNativeHandle(dev)`).

For SYCL devices, this does not work, as `getNativeHandle(dev)` returns a `pair(context, device)` instead of an index.

The CMS code would still benefit from a linear index, because it enumerates all the devices on each platform at the application startup, stores them in a `vector<Dev>` per platform, and refers to them by index.
And, for back-ends like SYCL where `getNativeHandle(dev)` does not return an index, we could implement some kind of lookup to go from the device object (or rather its `m_impl.get()` pointer) to the index in the `vector<Dev>`.

Is thins something that would make sense to have more generally in Alpaka ?

I guess to work it would also need to introduce a centrally managed list of devices for each back-end.

**Comment by psychocoderHPC** [comment]
Currently, alpaka is stateless besides a few `static` objects. If I read this issue and remember the discussions about the static objects in alpaka the question is IMO does alpaka should provide something like a "driver".

**Comment by bernhardmgruber** [comment]
I understand your use case, but I don't see this nicely fitting into alpaka. But people can convince me otherwise.

> For SYCL devices, this does not work, as getNativeHandle(dev) returns a pair(context, device) instead of an index.

Which is fine, since `getNativeHandle(dev)` should return you the "native handle" and not an index. The fact that CUDA/HIP return an index is IMO by accident and not by design.

> I guess to work it would also need to introduce a centrally managed list of devices for each back-end.

Which is what I would refrain from. 

> The CMS code would still benefit from a linear index, because it enumerates all the devices on each platform at the application startup, stores them in a vector<Dev> per platform, and refers to them by index.

And I think this is an alright solution. So the ownership of the device vectors stays in the client application and is not imposed on alpaka.

**Comment by fwyzard** [comment]
I _almost_ agree - but the fact that the SYCL platforms support `getDevByIdx()` breaks the model for me.

Either the SYCL devices are "unordered", and then `getDevByIdx()` should not be supported.
Or the SYCL devices have an intrinsic order, and the there should be a way to get the underlying index from the SYCL device object.

**Comment by bernhardmgruber** [comment]
Right, that's a fair point. This sounds to me like we may want to have a function to get the index for a device. We may not need to store the devices for that. If that function is not required to be fast, we could query all devices of the platform and then just find the positon of the current device in that list. Alternatively, each device could just store its index when it is created. And if we know that the native handle is a zero-based index, then we could just use that.

**Comment by j-stephan** [comment]
Since alpaka started out as a layer above CUDA it inherited many of CUDA's designs. This is IMO another case where we should rethink this approach. For the SYCL back-end we internally query the underlying SYCL platfom for the list of devices attached to it which gives us a `std::vector<sycl::device>`. When `alpaka::getDevByIdx` is called we return the `sycl::device` (or rather the `alpaka::DevSycl` encapsulating it) that corresponds to the given index in the vector.

> I guess to work it would also need to introduce a centrally managed list of devices for each back-end.

IMO a cleaner approach would be to copy SYCL's behaviour here. Let alpaka hand out a `std::vector<alpaka::Dev>` for each alpaka platform and remove `getDevByIdx`. The latter can easily be done by the user code if all platforms return a list of attached devices.

**Comment by fwyzard** [comment]
> Let alpaka hand out a `std::vector<alpaka::Dev>` for each alpaka platform and remove `getDevByIdx`. The latter can easily be done by the user code if all platforms return a list of attached devices.

Advantage: the _device_ objects can be created once and for all, which avoids the problem of creating multiple _device_ objects that refer to the same device through repeated calls to `getDevByIdx()`.

Possible problem: this approach creates a _device_ object for all devices, even those the user may never request. This could be somewhat inefficient, or confusing, or even break some very strange corner cases (e.g. a machine with different kind of GPUs).

**Comment by fwyzard** [comment]
That said, that's exactly what we do in the CMS code, both the standalone one and the full framework.

**Comment by fwyzard** [comment]
a compromise could be to initialise the devices and build the corresponding _device_ objects only when the user requests them, and internally store them in something like an `std::vector<std::optional<Device>>` .

**Comment by bernhardmgruber** [comment]
This should be easier to implement since platforms are objects now. Each platform could now just create a vector of devices, `getDevByIdx(platform, i)` just returns the `i`th device. In turn , we could then query the `i` for a device.

**Comment by fwyzard** [comment]
> Each platform could now just create a vector of devices, `getDevByIdx(platform, i)` just returns the `i`th device. In turn , we could then query the `i` for a device.

👍🏻

**Comment by fwyzard** [comment]
If we go along this route, my suggestion would be to create the device objects lazily, the first time time are requested:
  - `alpaka::getDevs(platform)` creates all device objects
  - `alpaka::getDevByIdx(platform, i)` creates only the _i_&nbsp;th object

**Comment by bernhardmgruber** [comment]
> If we go along this route, my suggestion would be to create the device objects lazily, the first time time are requested:
> 
> * `alpaka::getDevs(platform)` creates all device objects
> * `alpaka::getDevByIdx(platform, i)` creates only the _i_ th object

Sounds good to me.
